### PR TITLE
add type for extended patches

### PR DESCRIPTION
**Pull request by Liujingfang1**
Add type for patches according to [KEP](https://github.com/kubernetes/enhancements/pull/897),
which can be applied to multiple objects.

Also include the request for inlining patches. https://github.com/kubernetes-sigs/kustomize/issues/1157

### CODE REVIEW

File: pkg/types/kustomization.go [hunk 1]
```diff
@@ -376,3 +381,39 @@ type Replica struct {
...
+	Path string `json:"path,omitempty" yaml:"path,omitempty"`
+
+	// Patch is the content of a patch.
+	Patch string `json:"patch,omitempty" yaml:"patch,omitempty"`
+
+	// Target points to the resources that the patch is applied to
+	Target Target `json:"target,omitempty" yaml:"target,omitempty"`
+
+	// Type is one of `StrategicMergePatch` or `JsonPatch`.
+	// optional
```
Comment (monopole) [inline pkg/types/kustomization.go]: Type is one of "StrategicMerge" or "Json6902".
(Patch suffix redundant here)

I think this field should be required, so the kustomization is self-documenting.

File: pkg/types/kustomization.go [hunk 2]
```diff
@@ -376,3 +381,39 @@ type Replica struct {
...
+// Any resource that matches intersection of all conditions
+// is included in this set.
+type Targets struct {
+	gvk.Gvk   `json:",inline,omitempty" yaml:",inline,omitempty"`
+	Namespace string `json:"namespace,omitempty" yaml:"namespace,omitempty"`
+	Name      string `json:"name,omitempty" yaml:"name,omitempty"`
+
+	// MatchAnnotations is a map of key-value pairs.
+	// It matches with the resource annotations.
+	MatchAnnotation map[string]string `json:"matchAnnotation,omitempty" yaml:"matchAnnotation,omitempty"`
```
Comment (monopole) [inline pkg/types/kustomization.go]: Do MatchAnnotation and LabelSelector work the same way?
if so, use the same naming convention.

The term LabelSelector sound like it's going to support label selection _expressions_.  
https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#api

If that's the plan, the type should just be a bare string, not a map.

if it's a map, then how are all these matches going to be treated?  OR or AND?
Is there a NOT?
are label and annotation matching OR'd together or AND'd together?
If labels or annotations are specified and the Name is specified, is it an error?

File: pkg/types/kustomization.go (hunk 1 above)
Comment (Liujingfang1) [reply pkg/types/kustomization.go]: When we discussed this in the sig-cli meeting, the type of the patch can be determined from the content of the patch file.

File: pkg/types/kustomization.go (hunk 2 above)
Comment (Liujingfang1) [reply pkg/types/kustomization.go]: Thanks for the link. 
I think both the label selection expressions (a single string) and a map[string]string work.
The map will use AND for all key,value pairs to select.
The
One looks as
```
labelSelector: environment=production,tier=frontend
```
And the other looks as
```
labelSelector:
    environment: production
   tier: frontend
```
I think the later one looks better. We can change the field name to `targetLabel` to make it not confusing.

> are label and annotation matching OR'd together or AND'd together?
It's an AND of all requirements: annotations, labels, gvk, name, namespace

> If labels or annotations are specified and the Name is specified, is it an error?
Not necessary. The name can be a regexp.

File: pkg/types/kustomization.go (hunk 1 above)
Comment (monopole) [reply pkg/types/kustomization.go]: Good idea, just detect it.  Drop the Type field.

File: pkg/types/kustomization.go (hunk 2 above)
Comment (monopole) [reply pkg/types/kustomization.go]: IMO, use a string.  It's future proof.  You can introduce OR, NOT etc. later.  The map structure locks you into AND only (or OR only).   There's a reason why k8s does it that way.  Plus there's probably a  library somewhere that could be reused to do the parsing.

Field comments should mention where regexp is allows.

Comment (sethpollack) [reply pkg/types/kustomization.go]: maybe use this package? https://github.com/kubernetes/apimachinery/blob/master/pkg/labels/labels.go

Comment (Liujingfang1) [reply pkg/types/kustomization.go]: OK, then I'll change it to a string

Comment (Liujingfang1) [reply pkg/types/kustomization.go]: @monopole @sethpollack  How is the name `MatchAnnotation`? What about `AnnotationSelector`?

Comment (monopole) [reply pkg/types/kustomization.go]: voting "Selector" for both, if they behave like the existing LabelSelector

Comment (Liujingfang1) [reply pkg/types/kustomization.go]: updated to `AnnotationSelector` and `LabelSelector`.

**Pull request merged**